File: src/kx/graph.py
from kubernetes import client
from rich.tree import Tree

from kx.k8s import load_config
from kx.kinds import Kind


def build_tree(kind: str, name: str, namespace: str) -> Tree:
    root, _ = _build(kind, name, namespace, indexed=False)
    return root


def build_indexed_tree(
    kind: str, name: str, namespace: str
) -> tuple[Tree, list[tuple[str, str]]]:
    return _build(kind, name, namespace, indexed=True)


def build_namespace_tree(namespace: str) -> Tree:
    root, _ = _build_namespace(namespace, indexed=False)
    return root


def build_namespace_indexed_tree(
    namespace: str,
) -> tuple[Tree, list[tuple[str, str]]]:
    return _build_namespace(namespace, indexed=True)


def _build(
    kind: str, name: str, namespace: str, indexed: bool
) -> tuple[Tree, list[tuple[str, str]]]:
    load_config()
    resources: list[tuple[str, str]] = [(name, kind)] if indexed else []
    label = f"[header]{kind}/{name}[/header]"
    if indexed:
        label = f"[muted]1[/muted] {label}"
    root = Tree(label)

    apps = client.AppsV1Api()
    core = client.CoreV1Api()
    batch = client.BatchV1Api()

    pods = core.list_namespaced_pod(namespace).items

    match kind:
        case Kind.Deployment:
            _tree_deployment(name, namespace, root, apps, pods, resources, indexed)
        case Kind.ReplicaSet:
            _tree_replica_set(name, namespace, root, apps, pods, resources, indexed)
        case Kind.StatefulSet:
            _tree_stateful_set(name, namespace, root, apps, pods, resources, indexed)
        case Kind.DaemonSet:
            _tree_daemon_set(name, namespace, root, apps, pods, resources, indexed)
        case Kind.CronJob:
            _tree_cron_job(name, namespace, root, batch, pods, resources, indexed)
        case Kind.Service:
            _tree_service(name, namespace, root, core, resources, indexed)
        case Kind.Pod:
            pod = core.read_namespaced_pod(name, namespace)
            _add_containers(pod, root)
        case _:
            root.add(f"[muted](no ownership graph for {kind})[/muted]")

    return root, resources


# Prefix + semantic color for each node kind in a namespace forest. Controllers
# render as roots with their full kind name; owned resources use the short
# prefixes the single-resource tree already uses (rs/job/pod).
_NODE_STYLE: dict[Kind, tuple[str, str]] = {
    Kind.Deployment: ("Deployment", "header"),
    Kind.StatefulSet: ("StatefulSet", "header"),
    Kind.DaemonSet: ("DaemonSet", "header"),
    Kind.CronJob: ("CronJob", "header"),
    Kind.Job: ("job", "accent"),
    Kind.ReplicaSet: ("rs", "accent"),
    Kind.Pod: ("pod", "body"),
}

# Stable order roots appear under the Namespace node.
_ROOT_ORDER: list[Kind] = [
    Kind.Deployment,
    Kind.StatefulSet,
    Kind.DaemonSet,
    Kind.CronJob,
    Kind.Job,
    Kind.ReplicaSet,
    Kind.Pod,
]


def _build_namespace(
    namespace: str, indexed: bool
) -> tuple[Tree, list[tuple[str, str]]]:
    """Full ownership forest for a namespace: every workload controller as a
    root with its owned resources beneath, plus orphan Jobs/ReplicaSets and bare
    Pods so nothing is hidden. Each pod appears exactly once. Unlike `_build`,
    the Namespace root node is not indexed — children number from 1."""
    load_config()
    apps = client.AppsV1Api()
    core = client.CoreV1Api()
    batch = client.BatchV1Api()

    deployments = apps.list_namespaced_deployment(namespace).items
    replica_sets = apps.list_namespaced_replica_set(namespace).items
    stateful_sets = apps.list_namespaced_stateful_set(namespace).items
    daemon_sets = apps.list_namespaced_daemon_set(namespace).items
    cron_jobs = batch.list_namespaced_cron_job(namespace).items
    jobs = batch.list_namespaced_job(namespace).items
    pods = core.list_namespaced_pod(namespace).items

    # Resources that can be owned by something else in the namespace, indexed by
    # each owner uid so a parent finds its children in one pass.
    child_kinds = (
        [(rs, Kind.ReplicaSet) for rs in replica_sets]
        + [(job, Kind.Job) for job in jobs]
        + [(pod, Kind.Pod) for pod in pods]
    )
    children_by_owner: dict[str, list[tuple[object, Kind]]] = {}
    for obj, kind in child_kinds:
        for ref in obj.metadata.owner_references or []:
            children_by_owner.setdefault(ref.uid, []).append((obj, kind))

    present_uids = {
        obj.metadata.uid
        for obj in (
            *deployments,
            *replica_sets,
            *stateful_sets,
            *daemon_sets,
            *cron_jobs,
            *jobs,
            *pods,
        )
    }

    # Controllers are always roots; an owned child (rs/job/pod) is a root only
    # when orphaned — its owner isn't among the namespace's collected objects.
    roots: list[tuple[object, Kind]] = [
        *[(d, Kind.Deployment) for d in deployments],
        *[(s, Kind.StatefulSet) for s in stateful_sets],
        *[(d, Kind.DaemonSet) for d in daemon_sets],
        *[(c, Kind.CronJob) for c in cron_jobs],
    ]
    for obj, kind in child_kinds:
        refs = obj.metadata.owner_references or []
        if not any(ref.uid in present_uids for ref in refs):
            roots.append((obj, kind))

    order = {kind: position for position, kind in enumerate(_ROOT_ORDER)}
    roots.sort(key=lambda pair: (order[pair[1]], pair[0].metadata.name))

    resources: list[tuple[str, str]] = []
    root = Tree(f"[header]Namespace/{namespace}[/header]")
    if not roots:
        root.add("[muted](no workloads)[/muted]")
        return root, resources

    for obj, kind in roots:
        _render(obj, kind, root, children_by_owner, resources, indexed)
    return root, resources


def _render(obj, kind, parent, children_by_owner, resources, indexed):
    """Add `obj` under `parent`, then recurse into its owned children; a Pod is
    a leaf whose containers are listed directly."""
    prefix, color = _NODE_STYLE[kind]
    node = _add_node(parent, resources, indexed, color, prefix, obj.metadata.name, kind)
    if kind == Kind.Pod:
        _add_containers(obj, node)
        return
    for child, child_kind in children_by_owner.get(obj.metadata.uid, []):
        _render(child, child_kind, node, children_by_owner, resources, indexed)


def _add_node(parent, resources, indexed, color, prefix, name, kind):
    """Add a labelled child node; when indexed, prepend its 1-based index and
    record `(name, kind)` in `resources`."""
    label = f"[{color}]{prefix}/{name}[/{color}]"
    if indexed:
        label = f"[muted]{len(resources) + 1}[/muted] {label}"
        resources.append((name, kind))
    return parent.add(label)


def _tree_deployment(name, namespace, node, apps, pods, resources, indexed):
    deploy = apps.read_namespaced_deployment(name, namespace)
    uid = deploy.metadata.uid
    replica_sets = [
        rs
        for rs in apps.list_namespaced_replica_set(namespace).items
        if _owned_by(rs, uid)
    ]
    for rs in replica_sets:
        rs_node = _add_node(
            node,
            resources,
            indexed,
            "accent",
            "rs",
            rs.metadata.name,
            Kind.ReplicaSet,
        )
        _add_pods_for_owner(rs.metadata.uid, pods, rs_node, resources, indexed)


def _tree_replica_set(name, namespace, node, apps, pods, resources, indexed):
    rs = apps.read_namespaced_replica_set(name, namespace)
    _add_pods_for_owner(rs.metadata.uid, pods, node, resources, indexed)


def _tree_stateful_set(name, namespace, node, apps, pods, resources, indexed):
    sts = apps.read_namespaced_stateful_set(name, namespace)
    _add_pods_for_owner(sts.metadata.uid, pods, node, resources, indexed)


def _tree_daemon_set(name, namespace, node, apps, pods, resources, indexed):
    ds = apps.read_namespaced_daemon_set(name, namespace)
    _add_pods_for_owner(ds.metadata.uid, pods, node, resources, indexed)


def _tree_cron_job(name, namespace, node, batch, pods, resources, indexed):
    cj = batch.read_namespaced_cron_job(name, namespace)
    uid = cj.metadata.uid
    jobs = [
        job for job in batch.list_namespaced_job(namespace).items if _owned_by(job, uid)
    ]
    for job in jobs:
        job_node = _add_node(
            node, resources, indexed, "accent", "job", job.metadata.name, Kind.Job
        )
        _add_pods_for_owner(job.metadata.uid, pods, job_node, resources, indexed)


def _tree_service(name, namespace, node, core, resources, indexed):
    svc = core.read_namespaced_service(name, namespace)
    selector = svc.spec.selector
    if not selector:
        node.add("[muted](no selector)[/muted]")
        return
    label_selector = ",".join(f"{k}={v}" for k, v in selector.items())
    pods = core.list_namespaced_pod(namespace, label_selector=label_selector).items
    if not pods:
        node.add("[muted](no matching pods)[/muted]")
        return
    for pod in pods:
        pod_node = _add_node(
            node, resources, indexed, "body", "pod", pod.metadata.name, Kind.Pod
        )
        _add_containers(pod, pod_node)


def _add_pods_for_owner(owner_uid, pods, parent_node, resources, indexed):
    owned = [pod for pod in pods if _owned_by(pod, owner_uid)]
    for pod in owned:
        pod_node = _add_node(
            parent_node,
            resources,
            indexed,
            "body",
            "pod",
            pod.metadata.name,
            Kind.Pod,
        )
        _add_containers(pod, pod_node)


def _add_containers(pod, parent_node):
    for container in pod.spec.containers:
        parent_node.add(f"[muted]container: {container.name}[/muted]")


def _owned_by(resource, uid: str) -> bool:
    refs = resource.metadata.owner_references or []
    return any(ref.uid == uid for ref in refs)


def most_recent_job(owner_uid: str, jobs):
    """The most recently-created Job owned by `owner_uid` (a CronJob's uid),
    by `metadata.creation_timestamp`; None if it has never run. Shared by
    resolve_workload_pods and DiagnosticsService — CronJob health and pods
    are both scoped to the latest run, not the full retained history."""
    owned = [job for job in jobs if _owned_by(job, owner_uid)]
    if not owned:
        return None
    return max(owned, key=lambda job: job.metadata.creation_timestamp)


def resolve_workload_pods(
    kind: str, name: str, namespace: str, apps, core, batch=None
) -> list:
    """Resolve the pods belonging to a workload via ownership references.

    Deployment is a two-hop walk (Deployment → owned ReplicaSets → owned pods,
    which includes surge/old pods mid-rollout); StatefulSet/DaemonSet/Job own
    their pods directly; Pod resolves to itself. Reuses `_owned_by` and fetches
    pods once per namespace, filtering client-side (mirrors the tree
    builders). `batch` is required for Job, unused otherwise."""
    if kind == Kind.Pod:
        return [core.read_namespaced_pod(name, namespace)]
    if kind == Kind.Service:
        svc = core.read_namespaced_service(name, namespace)
        selector = svc.spec.selector
        if not selector:
            return []
        label_selector = ",".join(f"{k}={v}" for k, v in selector.items())
        return core.list_namespaced_pod(namespace, label_selector=label_selector).items

    pods = core.list_namespaced_pod(namespace).items
    match kind:
        case Kind.Deployment:
            deploy = apps.read_namespaced_deployment(name, namespace)
            rs_uids = {
                rs.metadata.uid
                for rs in apps.list_namespaced_replica_set(namespace).items
                if _owned_by(rs, deploy.metadata.uid)
            }
            return [pod for pod in pods if any(_owned_by(pod, uid) for uid in rs_uids)]
        case Kind.StatefulSet:
            sts = apps.read_namespaced_stateful_set(name, namespace)
            return [pod for pod in pods if _owned_by(pod, sts.metadata.uid)]
        case Kind.DaemonSet:
            ds = apps.read_namespaced_daemon_set(name, namespace)
            return [pod for pod in pods if _owned_by(pod, ds.metadata.uid)]
        case Kind.Job:
            job = batch.read_namespaced_job(name, namespace)
            return [pod for pod in pods if _owned_by(pod, job.metadata.uid)]
        case Kind.CronJob:
            cj = batch.read_namespaced_cron_job(name, namespace)
            recent = most_recent_job(
                cj.metadata.uid, batch.list_namespaced_job(namespace).items
            )
            if recent is None:
                return []
            return [pod for pod in pods if _owned_by(pod, recent.metadata.uid)]
        case _:
            return []

File: internal/graph/graph.go
// Package graph walks Kubernetes ownership references to build the tree kx
// renders.
//
// This is the one place kx needs the API server rather than kubectl: ownership
// references are not in kubectl's table output, and reconstructing them from
// `-o json` per resource would be many more round trips than listing each kind
// once.
package graph

import (
	"context"

	batchv1 "k8s.io/api/batch/v1"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/client-go/kubernetes"

	"github.com/jzills/kx/internal/kinds"
	"github.com/jzills/kx/internal/render"
	"github.com/jzills/kx/internal/theme"
)

// Resource is a node recorded for indexing, so `kx tree --index` can save the
// tree as state and later commands can resolve those indexes.
type Resource struct {
	Name string
	Kind kinds.Kind
}

// Builder walks ownership references against a Kubernetes API client.
type Builder struct {
	Client kubernetes.Interface
}

// nodeStyle gives the prefix and semantic style for each kind in a tree.
// Controllers render with their full kind name; owned resources use the short
// prefixes (rs/job/pod) the single-resource tree uses.
var nodeStyle = map[kinds.Kind]struct{ Prefix, Style string }{
	kinds.Deployment:  {"Deployment", theme.Header},
	kinds.StatefulSet: {"StatefulSet", theme.Header},
	kinds.DaemonSet:   {"DaemonSet", theme.Header},
	kinds.CronJob:     {"CronJob", theme.Header},
	kinds.Job:         {"job", theme.Accent},
	kinds.ReplicaSet:  {"rs", theme.Accent},
	kinds.Pod:         {"pod", theme.Body},
}

// rootOrder is the order roots appear under the Namespace node.
var rootOrder = []kinds.Kind{
	kinds.Deployment,
	kinds.StatefulSet,
	kinds.DaemonSet,
	kinds.CronJob,
	kinds.Job,
	kinds.ReplicaSet,
	kinds.Pod,
}

// collector accumulates the indexed resources as the walk adds nodes.
type collector struct {
	indexed   bool
	resources []Resource
}

// add appends a labelled child, numbering it and recording it when the tree is
// being indexed.
func (c *collector) add(parent *render.Node, style, prefix, name string, kind kinds.Kind) *render.Node {
	label := prefix + "/" + name
	if !c.indexed {
		return parent.Add(label, style)
	}
	c.resources = append(c.resources, Resource{Name: name, Kind: kind})
	return parent.AddIndexed(label, style, len(c.resources))
}

func ownedBy(meta metav1.ObjectMeta, uid types.UID) bool {
	for _, ref := range meta.OwnerReferences {
		if ref.UID == uid {
			return true
		}
	}
	return false
}

// BuildResource graphs a single resource's ownership tree.
func (b Builder) BuildResource(
	ctx context.Context, kind kinds.Kind, name, namespace string, indexed bool,
) (*render.Node, []Resource, error) {
	c := &collector{indexed: indexed}

	root := &render.Node{Label: string(kind) + "/" + name, Style: theme.Header}
	if indexed {
		// The root is itself indexable, and numbers first.
		c.resources = append(c.resources, Resource{Name: name, Kind: kind})
		root.Index = 1
	}

	// Pods are listed once and filtered client-side, mirroring how the
	// namespace forest works: one call rather than one per owner.
	pods, err := b.pods(ctx, namespace)
	if err != nil {
		return nil, nil, err
	}

	switch kind {
	case kinds.Deployment:
		err = b.treeDeployment(ctx, name, namespace, root, pods, c)
	case kinds.ReplicaSet:
		err = b.treeOwnedPods(ctx, kinds.ReplicaSet, name, namespace, root, pods, c)
	case kinds.StatefulSet:
		err = b.treeOwnedPods(ctx, kinds.StatefulSet, name, namespace, root, pods, c)
	case kinds.DaemonSet:
		err = b.treeOwnedPods(ctx, kinds.DaemonSet, name, namespace, root, pods, c)
	case kinds.CronJob:
		err = b.treeCronJob(ctx, name, namespace, root, pods, c)
	case kinds.Service:
		err = b.treeService(ctx, name, namespace, root, c)
	case kinds.Pod:
		pod, podErr := b.Client.CoreV1().Pods(namespace).Get(ctx, name, metav1.GetOptions{})
		if podErr != nil {
			return nil, nil, podErr
		}
		addContainers(pod, root)
	default:
		root.Add("(no ownership graph for "+string(kind)+")", theme.Muted)
	}
	if err != nil {
		return nil, nil, err
	}
	return root, c.resources, nil
}

func (b Builder) pods(ctx context.Context, namespace string) ([]corev1.Pod, error) {
	list, err := b.Client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, err
	}
	return list.Items, nil
}

func (b Builder) treeDeployment(
	ctx context.Context, name, namespace string, node *render.Node,
	pods []corev1.Pod, c *collector,
) error {
	deployment, err := b.Client.AppsV1().Deployments(namespace).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return err
	}
	replicaSets, err := b.Client.AppsV1().ReplicaSets(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return err
	}
	// A Deployment mid-rollout owns several ReplicaSets; all of them are shown
	// so surge and old pods are visible rather than silently dropped.
	for i := range replicaSets.Items {
		rs := &replicaSets.Items[i]
		if !ownedBy(rs.ObjectMeta, deployment.UID) {
			continue
		}
		rsNode := c.add(node, theme.Accent, "rs", rs.Name, kinds.ReplicaSet)
		addPodsForOwner(rs.UID, pods, rsNode, c)
	}
	return nil
}

// treeOwnedPods handles the kinds that own their pods directly.
func (b Builder) treeOwnedPods(
	ctx context.Context, kind kinds.Kind, name, namespace string,
	node *render.Node, pods []corev1.Pod, c *collector,
) error {
	uid, err := b.workloadUID(ctx, kind, name, namespace)
	if err != nil {
		return err
	}
	addPodsForOwner(uid, pods, node, c)
	return nil
}

func (b Builder) workloadUID(
	ctx context.Context, kind kinds.Kind, name, namespace string,
) (types.UID, error) {
	get := metav1.GetOptions{}
	switch kind {
	case kinds.ReplicaSet:
		object, err := b.Client.AppsV1().ReplicaSets(namespace).Get(ctx, name, get)
		if err != nil {
			return "", err
		}
		return object.UID, nil
	case kinds.StatefulSet:
		object, err := b.Client.AppsV1().StatefulSets(namespace).Get(ctx, name, get)
		if err != nil {
			return "", err
		}
		return object.UID, nil
	case kinds.DaemonSet:
		object, err := b.Client.AppsV1().DaemonSets(namespace).Get(ctx, name, get)
		if err != nil {
			return "", err
		}
		return object.UID, nil
	case kinds.Job:
		object, err := b.Client.BatchV1().Jobs(namespace).Get(ctx, name, get)
		if err != nil {
			return "", err
		}
		return object.UID, nil
	default:
		return "", nil
	}
}

func (b Builder) treeCronJob(
	ctx context.Context, name, namespace string, node *render.Node,
	pods []corev1.Pod, c *collector,
) error {
	cronJob, err := b.Client.BatchV1().CronJobs(namespace).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return err
	}
	jobs, err := b.Client.BatchV1().Jobs(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return err
	}
	for i := range jobs.Items {
		job := &jobs.Items[i]
		if !ownedBy(job.ObjectMeta, cronJob.UID) {
			continue
		}
		jobNode := c.add(node, theme.Accent, "job", job.Name, kinds.Job)
		addPodsForOwner(job.UID, pods, jobNode, c)
	}
	return nil
}

func (b Builder) treeService(
	ctx context.Context, name, namespace string, node *render.Node, c *collector,
) error {
	service, err := b.Client.CoreV1().Services(namespace).Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return err
	}
	if len(service.Spec.Selector) == 0 {
		// A headless or externalName Service selects nothing; say so rather
		// than rendering an empty branch.
		node.Add("(no selector)", theme.Muted)
		return nil
	}
	list, err := b.Client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{
		LabelSelector: formatSelector(service.Spec.Selector),
	})
	if err != nil {
		return err
	}
	if len(list.Items) == 0 {
		node.Add("(no matching pods)", theme.Muted)
		return nil
	}
	for i := range list.Items {
		pod := &list.Items[i]
		podNode := c.add(node, theme.Body, "pod", pod.Name, kinds.Pod)
		addContainers(pod, podNode)
	}
	return nil
}

func addPodsForOwner(uid types.UID, pods []corev1.Pod, parent *render.Node, c *collector) {
	for i := range pods {
		pod := &pods[i]
		if !ownedBy(pod.ObjectMeta, uid) {
			continue
		}
		podNode := c.add(parent, theme.Body, "pod", pod.Name, kinds.Pod)
		addContainers(pod, podNode)
	}
}

func addContainers(pod *corev1.Pod, parent *render.Node) {
	for _, container := range pod.Spec.Containers {
		parent.Add("container: "+container.Name, theme.Muted)
	}
}

// MostRecentJob returns the most recently created Job owned by uid (a CronJob's
// uid), or nil if it has never run.
//
// CronJob health and pods are both scoped to the latest run rather than the
// full retained history, so this is shared with the diagnostics.
func MostRecentJob(uid types.UID, jobs []batchv1.Job) *batchv1.Job {
	var recent *batchv1.Job
	for i := range jobs {
		job := &jobs[i]
		if !ownedBy(job.ObjectMeta, uid) {
			continue
		}
		if recent == nil || job.CreationTimestamp.Time.After(recent.CreationTimestamp.Time) {
			recent = job
		}
	}
	return recent
}

// ownerRef is one entry in the namespace forest's owner index.
type ownerRef struct {
	object metav1.Object
	kind   kinds.Kind
	// containers is set for pods, which render their containers as leaves.
	pod *corev1.Pod
}

// BuildNamespace graphs the whole ownership forest for a namespace: every
// workload controller as a root with its owned resources beneath, plus orphaned
// Jobs, ReplicaSets and bare Pods so nothing is hidden. Each pod appears
// exactly once.
//
// Unlike BuildResource, the Namespace root is not indexed — children number
// from 1.
func (b Builder) BuildNamespace(
	ctx context.Context, namespace string, indexed bool,
) (*render.Node, []Resource, error) {
	deployments, err := b.Client.AppsV1().Deployments(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}
	replicaSets, err := b.Client.AppsV1().ReplicaSets(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}
	statefulSets, err := b.Client.AppsV1().StatefulSets(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}
	daemonSets, err := b.Client.AppsV1().DaemonSets(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}
	cronJobs, err := b.Client.BatchV1().CronJobs(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}
	jobs, err := b.Client.BatchV1().Jobs(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}
	pods, err := b.Client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, nil, err
	}

	// Resources that something else in the namespace can own.
	var children []ownerRef
	for i := range replicaSets.Items {
		children = append(children, ownerRef{object: &replicaSets.Items[i], kind: kinds.ReplicaSet})
	}
	for i := range jobs.Items {
		children = append(children, ownerRef{object: &jobs.Items[i], kind: kinds.Job})
	}
	for i := range pods.Items {
		children = append(children, ownerRef{
			object: &pods.Items[i], kind: kinds.Pod, pod: &pods.Items[i],
		})
	}

	// Indexed by owner uid so a parent finds its children in one pass.
	childrenByOwner := map[types.UID][]ownerRef{}
	for _, child := range children {
		for _, ref := range child.object.GetOwnerReferences() {
			childrenByOwner[ref.UID] = append(childrenByOwner[ref.UID], child)
		}
	}

	present := map[types.UID]bool{}
	markPresent := func(uid types.UID) { present[uid] = true }
	for i := range deployments.Items {
		markPresent(deployments.Items[i].UID)
	}
	for i := range replicaSets.Items {
		markPresent(replicaSets.Items[i].UID)
	}
	for i := range statefulSets.Items {
		markPresent(statefulSets.Items[i].UID)
	}
	for i := range daemonSets.Items {
		markPresent(daemonSets.Items[i].UID)
	}
	for i := range cronJobs.Items {
		markPresent(cronJobs.Items[i].UID)
	}
	for i := range jobs.Items {
		markPresent(jobs.Items[i].UID)
	}
	for i := range pods.Items {
		markPresent(pods.Items[i].UID)
	}

	// Controllers are always roots. An owned child is a root only when
	// orphaned — its owner isn't among the namespace's collected objects —
	// which is what keeps every pod visible exactly once.
	var roots []ownerRef
	for i := range deployments.Items {
		roots = append(roots, ownerRef{object: &deployments.Items[i], kind: kinds.Deployment})
	}
	for i := range statefulSets.Items {
		roots = append(roots, ownerRef{object: &statefulSets.Items[i], kind: kinds.StatefulSet})
	}
	for i := range daemonSets.Items {
		roots = append(roots, ownerRef{object: &daemonSets.Items[i], kind: kinds.DaemonSet})
	}
	for i := range cronJobs.Items {
		roots = append(roots, ownerRef{object: &cronJobs.Items[i], kind: kinds.CronJob})
	}
	for _, child := range children {
		owned := false
		for _, ref := range child.object.GetOwnerReferences() {
			if present[ref.UID] {
				owned = true
				break
			}
		}
		if !owned {
			roots = append(roots, child)
		}
	}

	order := map[kinds.Kind]int{}
	for position, kind := range rootOrder {
		order[kind] = position
	}
	sortRoots(roots, order)

	c := &collector{indexed: indexed}
	root := &render.Node{Label: "Namespace/" + namespace, Style: theme.Header}
	if len(roots) == 0 {
		root.Add("(no workloads)", theme.Muted)
		return root, c.resources, nil
	}
	for _, entry := range roots {
		renderNode(entry, root, childrenByOwner, c)
	}
	return root, c.resources, nil
}

// renderNode adds an object under parent, then recurses into what it owns. A
// Pod is a leaf whose containers are listed directly.
func renderNode(entry ownerRef, parent *render.Node, childrenByOwner map[types.UID][]ownerRef, c *collector) {
	style := nodeStyle[entry.kind]
	node := c.add(parent, style.Style, style.Prefix, entry.object.GetName(), entry.kind)
	if entry.kind == kinds.Pod {
		addContainers(entry.pod, node)
		return
	}
	for _, child := range childrenByOwner[entry.object.GetUID()] {
		renderNode(child, node, childrenByOwner, c)
	}
}
